trace-agent: Add regex support for filtering tags

verdel opened the pull request:
### What does this PR do?
In my opinion it's more flexible to use regex pattern matching instead of a direct comparison in filtering tags process.

For example datadog creates and adds additional span tags(`http.path_group` or `http.url_details.port`) to spans only after ingestion. So accodring to docs `...These tags cannot be used to drop traces at the Datadog Agent level...`

If we add regex pattern support to the tag filtering process, it will be possible for example to filter the incoming tracing stream based on the incomplete occurrence of a string in the `http.url` tag.

### Motivation
Resolves #13218

### Possible Drawbacks / Trade-offs
Regex-based filtering may be more resource-intensive compared to string comparison. But it should be noted that tag filtering only apply to the root span of a trace. Therefore this should not have a critical impact on overall performance.

verdel committed:
trace-agent: add regex support for filtering tags

This change add support of regex pattern for filtering tags based on
value of apm_config.filter_tags or environment variables
DD_APM_FILTER_TAGS_REQUIRE and DD_APM_FILTER_TAGS_REJECT.

Example configuration:

	apm_config:
		filter_tags:
			require: ["db:^sql[0-9]{1}$", "db.instance:^mysql$"]
			reject: ["outcome:^success$"]

This increase flexibility of the tags filtering process.

verdel committed:
rework unit/integration tests for filter_tags to support regex pattern

verdel committed:
update documentation for filter_tags

djmitche commented on pkg/config/config_template.yaml:
```diff
@@ -1141,13 +1141,13 @@ api_key:
   ## Note: Rules take into account the intersection of tags defined.
   #
   # filter_tags:
-  #     require: [<LIST_OF_KEY_VALUE_TAGS>]
-  #     reject: [<LIST_OF_KEY_VALUE_TAGS>]
```
Changing explicit matching to regex matching is a breaking change.  For example, `http.path_group` will now match more strings (such as `http_path_group`).

One option here is to use a different JSON structure for the two

```yaml
require: ['foo.bar', {"regex": "foo\..*\.bar"}]
```
Here, the new syntax is never valid for older agents, so there's no conflict. However, parsing this with the existing config library may be difficult.

verdel replied:
I'll try to think how to avoid this breaking change soon

verdel replied:
Maybe the right solution would be to add a new block(filter_tags_regex) to the configuration and corresponding env variables for reject and require?

djmitche replied:
That could make sense.  You'd need to be careful to define the interactions: spans that match (require OR require_regex) AND NOT (reject or reject_regex)

verdel replied:
I add new configuration option `filter_tags_regex`.

When `filter_tags` and `filter_tags_regex` are used at the same time, all rules are united for filtering.
In the case when rule in `filter_tags` and `filter_tags_regex` have the same key, rule from `filter_tags` have precendence over the `filter_tags_regex`.

djmitche commented:
(Note, I only looked at the config changes -- others may have comments on other parts of this PR!)

verdel committed:
trace-agent: add new config option filter_tags_regex

This change add support of regex pattern for filtering tags based on
value of apm_config.filter_tags_regex or environment variables
DD_APM_FILTER_TAGS_REGEX_REQUIRE and DD_APM_FILTER_TAGS_REGEX_REJECT.

Example configuration:

    apm_config:
        filter_tags_regex:
            require: ["db:^sql[0-9]{1}$", "db.instance:^mysql$"]
            reject: ["outcome:^success$"]

verdel committed:
add unit/integration tests for filter_tags_regex

verdel committed:
update documentation for filter_tags and add for filter_tags_regex

djmitche commented on pkg/config/config_template.yaml:
```diff
@@ -1144,11 +1144,24 @@ api_key:
   ##  * reject - list of key or key/value strings - traces with these tags are dropped by the Agent
   ## Note: Rules take into account the intersection of tags defined.
+  ##       When `filter_tags` and `filter_tags_regex` are used at the same time, all rules are united for filtering.
+  ##       In the case when rule in `filter_tags` and `filter_tags_regex` have the same key, rule from `filter_tags`
+  ##       have precendence over the `filter_tags_regex`.
```
For `filter_tags`, my understanding is that a string like `foo` appearing in either require or reject will match tags like `foo`, `foo:bar`, `foo:bing`, and a string like `foo:bar` will only match the tag `foo:bar`.  The existing comments above describing that behavior are less than clear, and perhaps some cleanup could be helpful :)

But, I'm not sure what that means for `filter_tags_regex`.  How does a regex have a key?  Are the strings in `<LIST_OF_KEY_VALUE_REGEX_TAGS>` of the form `<regex>:<regex>`?

verdel replied:
How do you think, may be better to write like this:
```
## @param filter_tags - object - optional
  ## Defines rules by which to filter traces based on tags.
  ##  * require - list of key or key/value strings - traces must have those tags in order to be sent to Datadog
  ##  * reject - list of key or key/value strings - traces with these tags are dropped by the Agent
  ## Please note that:
  ##   1. Rules take into account the intersection of tags defined.
  ##   2. When `filter_tags` and `filter_tags_regex` are used at the same time, all rules are united for filtering.
  ##      In the case when rule in `filter_tags` and `filter_tags_regex` have the same key, rule from `filter_tags`
  ##      have precendence over the `filter_tags_regex`.
  ##
  ##      In the case of the following configuration:
  ##        filter_tags:
  ##          require: ["foo:bar"]
  ##        filter_tags_regex:
  ##          require: ["foo:^bar[0-9]{1}$"]
  ##      resulting combined rule will match `foo:bar` and not `foo:bar1`
  #
  # filter_tags:
  #     require: [<LIST_OF_KEY_VALUE_TAGS>]
  #     reject: [<LIST_OF_KEY_VALUE_TAGS>]

  ## @param filter_tags_regex - object - optional
  ## Defines rules by which to filter traces based on tags with regex pattern for tag values.
  ##  * require - list of key or key/value regex pattern strings - traces must have those tags in order to be sent to Datadog
  ##  * reject - list of key or key/value regex pattern strings - traces with these tags are dropped by the Agent
  ## Note: Rules take into account the intersection of tags defined.
  #
  # filter_tags_regex:
  #     require: [<LIST_OF_KEY_VALUE_REGEX_TAGS>]    # e.g. ["<key>:<regex>"]
  #     reject: [<LIST_OF_KEY_VALUE_REGEX_TAGS>]     # e.g. ["<key>:<regex>"]
```

djmitche replied:
Much better!

verdel replied:
✅

djmitche commented on pkg/config/config_template.yaml:
```diff
@@ -1144,11 +1144,24 @@ api_key:
+  ##  * reject - list of key or key/value regex pattern strings - traces with these tags are dropped by the Agent
+  ## Note: Rules take into account the intersection of tags defined.
+  #
+  # filter_tags:
```
```suggestion
  # filter_tags_regex:
```

verdel committed:
fix typo in pkg/config/config_template.yaml

Co-authored-by: Dustin J. Mitchell <dustin@v.igoro.us>

verdel committed:
add more detailed information about filter_tags and filter_tags_regex to the documentation

hestonhoffman commented on pkg/config/config_template.yaml:
```diff
@@ -1143,12 +1143,33 @@ api_key:
+  ## Please note that:
+  ##   1. Rules take into account the intersection of tags defined.
+  ##   2. When `filter_tags` and `filter_tags_regex` are used at the same time, all rules are united for filtering.
+  ##      In the case when rule in `filter_tags` and `filter_tags_regex` have the same key, rule from `filter_tags`
```
```suggestion
  ##      In cases where rules in `filter_tags` and `filter_tags_regex` have the same key, the rule from `filter_tags`
```

verdel replied:
✅

hestonhoffman commented on pkg/config/config_template.yaml:
```diff
@@ -1143,12 +1143,33 @@ api_key:
+  ##   1. Rules take into account the intersection of tags defined.
+  ##   2. When `filter_tags` and `filter_tags_regex` are used at the same time, all rules are united for filtering.
+  ##      In the case when rule in `filter_tags` and `filter_tags_regex` have the same key, rule from `filter_tags`
+  ##      have precendence over the `filter_tags_regex`.
```
```suggestion
  ##      takes precedence over the rule from `filter_tags_regex`.
```

verdel replied:
✅

verdel replied:
✅

hestonhoffman commented on pkg/config/config_template.yaml:
```diff
@@ -1143,12 +1143,33 @@ api_key:
+  ##          require: ["foo:bar"]
+  ##        filter_tags_regex:
+  ##          require: ["foo:^bar[0-9]{1}$"]
+  ##      resulting combined rule will match `foo:bar` and not `foo:bar1`
```
```suggestion
  ##      The resulting combined rule will match `foo:bar` and not `foo:bar1`
```

verdel replied:
✅

hestonhoffman reviewed (approved):
Small edits, but docs LGTM

verdel committed:
add requested changes into pkg/config/config_template.yaml

verdel commented:
@hestonhoffman, thank you for your comments. I correct docs accordingly.

djmitche reviewed (approved):
Config looks good to me now.  I'll leave the remainder of the review to the APM team.

verdel commented:
@kacper-murzyn, maybe I can somehow help with the process of merging this PR? It has been open for 4 months.

verdel commented:
@djmitche, Hi. Maybe I can somehow help with the process of merging this PR? It has been open for 6 months. In my opinion the pr is stuck.

kacper-murzyn commented:
@verdel Sorry for missing this. @djmitche is no longer in Datadog, so we need someone else to help with this. @katiehockman this seems to be APM related, could you help maybe?

trevdor commented:
> Config looks good to me now. I'll leave the remainder of the review to the APM team.

Hi APM team. 👋🏻 
As a new Datadog user, can you offer guidance: 
Is it the Dd position that we should write a custom opentelemetry sampler instead of relying on datadog.yml? This feature seems like just what my team needs, but I get the sense from this PR lying stagnant that it's not the direction Dd officially wants to move.

katiehockman commented:
Apologies everyone for the delay in getting back to you here! It definitely wasn't my intention to ignore anyone, or insinuate that DD's official direction is one way or another. I just missed the email notification 😅 

We are looking into this, and I'll get back in touch. Thanks for the contribution here, and for all of the comments and questions.

katiehockman commented on pkg/trace/config/config.go:
```diff
@@ -385,6 +385,12 @@ type AgentConfig struct {
 	// RejectTags specifies a list of tags which must be absent on the root span in order for a trace to be accepted.
 	RejectTags []*Tag
 
+	// RequireTagsRegex specifies a list of tags which must be present on the root span in order for a trace to be accepted.
```
```suggestion
	// RequireTagsRegex specifies a list of regexp for tags which must be present on the root span in order for a trace to be accepted.
```

katiehockman replied:
(here and below)

verdel replied:
Fix

katiehockman commented on pkg/trace/agent/agent_test.go:
```diff
@@ -761,9 +763,81 @@ func TestFilteredByTags(t *testing.T) {
 			span:    pb.Span{Meta: map[string]string{"something": "else", "key-only": "but-also-value", "bad-key-only": "random"}},
 			drop:    true,
 		},
+		{
```
I love the new tests, thanks for doing that. I think it's worth giving each test a name that describes what it's checking. That would make it easier to review and make sure all test cases are covered, and would help with debugging if one of the tests starts failing in the future. If the name gets too long, even just a little comment for each test would be good.

verdel replied:
I tried to come up with suitable names for the tests. If you have any additional ideas on how to improve the titles, I will do it right away

katiehockman commented on pkg/config/config_template.yaml:
```diff
@@ -1143,12 +1143,33 @@ api_key:
+  ## Please note that:
+  ##   1. Rules take into account the intersection of tags defined.
+  ##   2. When `filter_tags` and `filter_tags_regex` are used at the same time, all rules are united for filtering.
+  ##      In cases where rules in `filter_tags` and `filter_tags_regex` have the same key, the rule from `filter_tags`
```
```suggestion
  ##      In cases where rules in `filter_tags` and `filter_tags_regex` match the same key, the rule from `filter_tags`
```

verdel replied:
Fix

katiehockman commented on pkg/config/config_template.yaml:
```diff
@@ -1143,12 +1143,33 @@ api_key:
+  ##      In cases where rules in `filter_tags` and `filter_tags_regex` have the same key, the rule from `filter_tags`
+  ##      takes precendence over the rule from `filter_tags_regex`.
+  ##
+  ##      In the case of the following configuration:
```
```suggestion
  ##      For example, in the case of the following configuration:
```

verdel replied:
Fix

katiehockman commented on pkg/config/config_template.yaml:
```diff
@@ -1143,12 +1143,33 @@ api_key:
+  ##          require: ["foo:bar"]
+  ##        filter_tags_regex:
+  ##          require: ["foo:^bar[0-9]{1}$"]
+  ##      The resulting combined rule will match `foo:bar` and not `foo:bar1`
```
Rather than saying what it will match, can you instead state the result. e.g. "With these rules, traces with a tag `<insert_tag>` will be dropped, and those with a `<insert_tag>` tag  will be kept" (or something similar)

verdel replied:
I described in a little more detail what the result will be as a result of filtering span with the specified rules

katiehockman commented on pkg/trace/agent/agent.go:
```diff
@@ -536,18 +536,29 @@ func traceContainsError(trace pb.Trace) bool {
 }
 
-func filteredByTags(root *pb.Span, require, reject []*config.Tag) bool {
+func filteredByTags(root *pb.Span, require, reject []*config.Tag, requireRegex, rejectRegex []*config.TagRegex) bool {
```
I think it's worth adding a benchmark here for spans with many tags. If it's significant (which it likely will be, considering it's regexp), then we may want to note in the docs that this could have performance implications.

verdel replied:
Unfortunately, I no longer participate in the project in which the datadog agent was used. Therefore, I have no way to assess how much the use of regexp for span filtering reduces the tracking performance.

katiehockman replied:
I just meant a [Go benchmark ](https://pkg.go.dev/testing#hdr-Benchmarks) for this function with and without `rejectRegex` and `requireRegex` set. I assume that running a regular expression on every tag of each root span could have performance implications. But hopefully that doesn't need to be explicitly stated in the release notes, and users will understand that without it being stated outright. Fortunately this is only run on the root span of a trace, not every single span, so that should make this a bit less painful. 

If you don't feel comfortable adding a benchmark or it's not feasible, just let me know!

verdel replied:
That's it, now I understand what benchmark we are talking about.
I added a benchmark.

Here are the Benchmark launch results:
```shell
BenchmarkFilteredByTags/FilteredByTags-10         	46732989	        25.03 ns/op
BenchmarkFilteredByTags/FilteredByRegexTags-10    	14361446	        83.38 ns/op
```

I also add information about performance degradation when using regex
patterns to the description of the `filter_tags_regex` parameter.

verdel replied:
@katiehockman, If it doesn't bother you, will you be able to write your opinion about the latest changes. After that we will be able to mark this conversation as resolve.

katiehockman replied:
Looks great to me, thanks. I'll resolve this one.

katiehockman commented on pkg/config/config_template.yaml:
```diff
@@ -1143,12 +1143,33 @@ api_key:
+  ## Defines rules by which to filter traces based on tags with regex pattern for tag values.
+  ##  * require - list of key or key/value regex pattern strings - traces must have those tags in order to be sent to Datadog
+  ##  * reject - list of key or key/value regex pattern strings - traces with these tags are dropped by the Agent
+  ## Note: Rules take into account the intersection of tags defined.
```
Since we're not copying al of the docs from `filtering_tags` into the `filtering_tags_regexp` docs, let's just call out that someone should look there for full details about usage.

verdel replied:
Added a reference to the description of the filtering_tags parameter

katiehockman commented on cmd/trace-agent/config/config.go:
```diff
@@ -326,6 +326,27 @@ func applyDatadogConfig(c *config.AgentConfig) error {
+		tags := coreconfig.Datadog.GetStringSlice("apm_config.filter_tags_regex.require")
+		for _, tag := range tags {
+			splitTag := splitTagRegex(tag)
+			if containsKey(c.RequireTags, splitTag.K) {
+				continue
+			}
```
```suggestion
			if containsKey(c.RequireTags, splitTag.K) {
			    // RequireTags already has this tag, so skip the regexp.
				continue
			}
```

katiehockman replied:
(here and below, just to make it easier for the next reader of the code 😄 )

verdel replied:
I added the specified comment to the code

katiehockman reviewed (commented):
Apologies again for the delay in reviewing this, and thanks for your contribution! It looks like there are some merge conflicts, so let us know if you need help resolving those. The code looks reasonable to me, I just have a few suggestions to improve the docs/comments mostly.

verdel committed:
rework unit tests for filter by tags

verdel committed:
rework documentation for filter_tags and filter_tags_regex

verdel committed:
add and fix comments in code

verdel commented:
@katiehockman, I'm very sorry, but I fixed merge conflict via rebase of my branch from upstream main. Only then I read that it is better to do merge and not do force push after opening PR.

verdel committed:
fix unit test TestFilteredByTags in trace agent pkg

verdel committed:
add benchmark to compare the filtering by tags with and without regex

Also add information about performance degradation when using regex
patterns to the description of the filter_tags_regex parameter

katiehockman commented on pkg/config/config_template.yaml:
```diff
@@ -1435,6 +1435,9 @@ api_key:
   ##  * require - list of key or key/value regex pattern strings - traces must have those tags in order to be sent to Datadog
   ##  * reject - list of key or key/value regex pattern strings - traces with these tags are dropped by the Agent
   ## Note: Rules take into account the intersection of tags defined.
+  ## Using regex patterns in the tag filtering reduces performance by about 3 times.
```
nit: I'd just say "Using regexp patterns for tag filtering can have performance implications, and be a bit slower than typical tag filtering without regexp. However, this regexp is only run on the root span of a trace, so should not have a critical impact on overall performance."

katiehockman replied:
I just went ahead and committed this directly so we can merge this PR :)

katiehockman committed:
Update config_template.yaml

verdel commented:
@katiehockman, will something else be required from me in order to merge this PR?

katiehockman commented:
@verdel nope just need to wait for CI to finish up, then I can go ahead and merge it for you. Thanks again for contributing this, and for your patience with the review.

verdel commented:
Thank you very much for your time and PR review

katiehockman commented:
@verdel do you mind trying to rebase? It looks like the CI is stuck, so I'm hoping a rebase (force pushing is fine if you have to) will help it get unstuck

verdel commented:
> @verdel do you mind trying to rebase? It looks like the CI is stuck, so I'm hoping a rebase (force pushing is fine if you have to) will help it get unstuck

I rebase my branch to upstream/main and force pushing commits again.

katiehockman commented:
Merged! Thanks again @verdel!
